### ISSUE TITLE
Logical module namespaces

### ISSUE THREAD

**Issue by dcleao** [jkrems/proposal-pkg-exports#42]
Statically typed languages such as Java or C# commonly define types in logical namespaces using special syntax (e.g. `package com.apple.quicktime`, in Java). Some of these languages enforce that each file has a single top-level type definition and that the file name and directory structure above it matches the logical namespace declared in the file's content. This rule helps avoid confusion.

In these languages, deployment of types as a cohesive whole, and their versioning, is a totally separate concern. 

In Java, multiple files are bundled in a JAR file. The JAR file has a name which does not necessarily match the logical namespaces defined by the files within. JAR files are then published in a certain repository. When stored in a Maven-like repository, these are identified by a Group name, an Artifact name and a Version, or GAV. Dependencies are declared in a maven project by referencing their GAV.
In .Net, similarly, multiple files are bundled in an Assembly file. Assembly files have a name and version which is unrelated with the contained logical namespaces.

In these languages, a single unit of dependency can contain types belonging to multiple top-level logical namespaces. Additionally, a single top-level logical namespace may be composed (contributed to) by multiple units of dependency.

Although not enforced, logical namespaces typically parallel domain names which are registered under global authorities, such as DNS.

---

The JavaScript language and ecosystem is different in many respects. JavaScript has no explicit concept of namespace, and types are not so strictly defined. In JavaScript there is a global object (and scope) in which "values" can be published. In JavaScript, functions are values. So, naturally, and traditionally, JavaScript files "exported" one or more values ("classes", functions, singleton objects) by publishing plain objects, accessible from the global object, forming a hierarchical namespace. Simply loading a file would export the values it defined. Together, all script files would build a tree of "values" accessible using dot notation from the top-level objects. Doing this was always clumsy. Partly, this was because there is no syntactic sugar to define such logical namespaces built with plain objects; it is always necessary to check if each segment of the namespace already exists, creating it if not. No JavaScript utility library became a standard to this effect. Also, in some browsers, the global scope and the global object were not true synonymous (they used to be, in previous versions of JavaScript) or there were quirks when defining global variables, explicitly or implicitly. Whatever, it never took off as a reliable way to publish exports of libraries. Perhaps, the main reason this export technique started being seen as such a bad practice was that complex pages started having the need to have multiple versions of the same library side-by-side and this is simply not possible with a single, global namespace (think `jquery`). Interestingly, in this respect .Net and Java also do not allow this (easily at least). Different versions of a logical namespace have to be loaded in separate `AppDomain`s and `ClassLoader`s , respectively.

Later came module formats for JavaScript, such as CommonJS and AMD, where modules could export one or more "values".

In CommonJS and Node.js, modules would be imported using a name which is anchored on the name of the package which contains them, thus tying the logical name to the deployment name. The package version is not included in the logical name. This simplicity does have the advantage of relying on the uniqueness of package names to ensure the uniqueness of namespaces, which are always rooted on the latter. However, this tends to leak non-logical organizational concerns to the names used to reference modules, and, implicitly, the "values" that they export (e.g. `import BarChart from "@my/lib/dist/es6/BarChart"`). As already said, above, this also impedes logical names to cross packages and packages to export more than one top-level logical name.

The AMD module format allows defining custom logical namespaces to reference modules which are totally unrelated to the URLs which expose them.

The forth-coming import maps standard, which one day will be supported by all web-browsers, also allows for this kind of logical namespace independence.

I believe that it is a goal of this project to provide a form to distribute import maps information across multiple projects, targeting either applications which are later made available in a web browser or in Node.js.

---

To allow this proposal to support packages using a logical namespace scheme, I propose a simple change: that the `exports` property _additionally_ allows for keys which do not start with `"./"`. When a bare key would be specified, it would not be appended to the package name but instead would form an independent logical namespace root.

The core package.
```json
{
  "name": "@my-org/core-lib",
  "version": "1.0.0",
  "exports": {
     "myOrg/util": "./util",
     "myOrg/components": "./components",
  }
}
```

In another package:
```json
{
  "name": "@my-org/stock-components",
  "version": "1.0.0",
  "exports": {
     "myOrg/components/Bar": "./components/Bar",
     "myOrg/components/Line": "./components/Line"
  }
}
```

In a JS file from a package which depends on the latter packages:
```js
import slice from "myOrg/util/slice";
import BarChart from "myOrg/components/Bar";

// ...
```

If I'm not mistaken, the "imports" property already allows creating internal, logical aliases with any desired bare name and so this proposal would just raise the restriction for the "exports" property.

**Comment by hybrist** [comment]
I think that explanation gives Java too much and JavaScript too little credit. Let me elaborate: When JavaScript moved to modules and packages (via npm/node), one important feature was that it didn't repeat a big mistake (imo) of Java: Conflating linking and library internals. Packages could contain hints for linking (their name) but the were just that. You could just put them into a different directory and everything just worked - the package now was available under a different name. Potentially even locally, only for a subset of files.

This proposal aims to provide additional hints for the binding/linking *without* removing that flexibility. Your example is already possible today in many tools, it's how `preact` can be used as a drop-in for `react`. It's also worth noting that the introduction of multiple package registries will most likely decouple `name` from the namespace to some degree. Hopefully using the mechanism most natural to JS - URLs. Allowing packages to rescope themselves brings with it all the pains of Java-style packages where there's a global namespace for everything, each class decides where it wants to be, and the only reason to have a local fork of an existing piece of code is to edit every single file and rescope it, hoping that it won't collide with somebody else's scopes.

I'm not sure this proposal is a good fit for your idea, I'd much rather see it as a proposal for `npm` and I believe it should always be driven from the consumer, never from the producer of the package (with the exception of light hints possibly).

**Comment by dcleao** [comment]
@jkrems:

I don't think I follow your arguments on Java (and many, many other programming languages).
I am not defending the specific aspects of how Java (or others) handles class loading, like class paths and such.

I am defending the flexibility that comes with being able to separate logical names from packaging names. Naturally, this flexibility comes at the cost of having to _secure_ two base names, one in a package registry and another in a logical namespace registry. In practice, the uniqueness of logical namespaces is managed informally, by using a logical prefix which is already centrally registered in another namespace registry, such as DNS. Even if someone uses "your" logical namespace in its own package, that is easily spotted and seen by anyone as incorrect and abusive. The only acceptable use cases for that — and note that this is not the main use case for which I am proposing this feature — would be to provide an alternate implementation of types exported by a package that you do not control (like replacing `react` with `preact` as you say). You could go further and try to impede someone from using "your" logical namespaces (if there was a way to register those), using some kind of signing system, but I don't think this is justified or necessary. That's how things work in most mainstream statically compiled languages and I don't consider this being a bad design, on the contrary. 

> It's also worth noting that the introduction of multiple package registries will most likely decouple name from the namespace to some degree. Hopefully using the mechanism most natural to JS - URLs.

Like you say, in the JavaScript world, with the advent of other package registries concurring with NPM (such as github), package names are no longer assured to refer to a single package definition and even organization — there is no central registry anymore. So, unfortunately, JavaScript lost the advantage it had on this front. Names you use to import modules no longer refer to a globally defined package — you have to add extra \@scope-to-registry mapping metadata to totally identify a package. As such, the proposed logical namespaces are on pair with the current package-based namespaces, in terms of uniqueness guarantees. Consistency comes from the _current set of installed packages_, in a certain evaluation environment, not colliding with each other (unique package names and unique logical export keys).

Also, I don't think that URLs are the most natural identification scheme for JS — URLs are kind of an implementation and installation detail. At most, a right-most part of a URL could be considered, ignoring arbitrary servers/hosts. But then, you've kind of arrived to the logical case again... (either rooted on package names or purely logical names).

> Allowing packages to rescope themselves brings with it all the pains of Java-style packages where there's a global namespace for everything, each class decides where it wants to be

> I believe it should always be driven from the consumer, never from the producer of the package (with the exception of light hints possibly).

What you consider "pains" can be seen as a feature. Isn't a _global namespace_ for everything what you have (or wished to have) with NPM package names? And, then, of course that each class is organized under some logical place by the library's author! When you place a module `c.js` and export it as `@a/b/c/d.js`, you are doing the exact same thing, organizing the contained types/values in some logical way determined by the publishing author. Unless the consumer of the package remaps these nested modules using `imports`, as proposed, he/she will have to "put up" with the publishing author's default (and healthy) organization.

So, we're only discussing about the choice of _root names_ in which the intra-package organization is exposed.

> I'm not sure this proposal is a good fit for your idea, I'd much rather see it as a proposal for npm

Both naming schemes are perfectly translatable to import maps, which is the ultimate/main target and base semantic model of your proposal, so how can this idea not fit?
Imo, your proposal should _not_ limit the choice of base names that can be exported and instead should leave that choice to library authors. Just like it does not for names declared in `imports`.

I fail to see how you think that this idea would make sense in the context of NPM? I am specifically trying to detach the logical, coding entities namespace from the naming of the package system... I'm trying to remove _packaging concerns_ from the source code of JS files...

I am not proposing something which _forces_ you to design your packages' namespaces the way I like them. Instead, I am proposing that you add the flexibility for each package author to choose the preferred naming scheme with which the functionality that they provide is consumed.

**Comment by dcleao** [comment]
I found a leak in my reasoning.
In my proposal, imports would be able to reference a _logical namespace of code entities_, separate of packaging and versioning concerns. It would be so independent that, some day later, we could change completely the packages providing/implementing such entities. For example, we could change the package providing the module `my/util/module.js` from `@my/util@2.0.0` to `@other/visual@1.0.0`. Names and versions of the before or after packages could be totally unrelated. This would be (or could be seen as) the advantage of the total packaging and versioning independence that I proposed. It makes even more sense (to me at least) and be more independent if we wouldn't need to specify the `.js` in the import statement.

My point with this exercise is that it becomes evident that for the switch of the _implementing_ package to work, there really exists an _implied_ version of the said logical coding entities. This version is not said anywhere, nor in the code, when importing, nor in the package.json. It is implied. There's some documentation page, somewhere, that says that the global concept `my/util/module.js`, version X (independent from that of either packages above), has a certain interface.

There lies the problem. Such a version actually exists, albeit not written, and it is not said anywhere, so it is not possible to validate that the actual packages, before or after, provide implementations for the said logical concepts.

This works to a certain extent, if the developer knows what he/she is doing. It has worked for many years, in many programming languages. When doing the switch, the developer takes special care to verify that the _after_ package does in fact implement the same version of the concepts that the _before_ package does.

Now, consider that you are creating an application that creates dashboards which are composed of multiple components, of arbitrary "types". When persisting a dashboard instance to disk, some specification would record the types of each of the components. The types would be represented by the name of their logical coding entity — as proposed, a name with no packaging or versioning concerns. At a later time, this file would be loaded and an implementation of the persisted type name would need to be found. Because the concept version was implied, it was not persisted at the time. When loading, there's no way to know if the version of the persisted concept is the same as is now available in the application. This may totally fail, when reading back any properties which were  persisted with the component, if the concept versions provided by the implementation packages are different.
So, in this case, we conclude that to do this right, we need to know and persist the version of the concept/component with the dashboard file. As such, the identifier of the type would now consist of a name and a version. The application code would have to have explicit knowledge of the concept versions of the components/types it was using. One way would be for the logical namespace to include the logical version, somewhere along the namespace segments.

So, to take this logical/packaging separation all the way, we'd need to also have explicit logical versions and move them around. This seems rather painful to do.

In fact, it seems, OSGi did exactly this, for Java. The exports of an OSGi bundle are versioned, logical names (versioned Java packages...); these versions are independent of the version of the bundle.

---

If, like in your current proposal, logical and packaging names are tied, names already have an implied version — that of the package on which names are always rooted. It just becomes tempting and far easier to rely on the package's version to version the contained concepts and assume that those concepts cannot "move" from one package to another; that they actually belong to a package. However, the mentioned disadvantages of this conflation are not worth (for me) the management of separate packaging and concept versions.

For more dynamic cases, like the dashboard example above, an API, having the knowledge of the import map, could allow "reflecting" on modules for knowing their package's name and version. A versioned, "canonical" module identifier could even be obtained (think old days' AMD `module.id`) — canonical because an import map can generally define multiple keys for a given URL.

> module URL => canonical, versioned module identifier.

Maybe Node.js could provide this functionality as a host specific property: `import.meta.canonicalId` (returning something like `@my/util@1.0.0/module.js`).
